### PR TITLE
Application Insights: CRUD for a new Instance

### PR DESCRIPTION
**Pull request by tombuildsstuff**
This is a Work-in-Progress designed to get some early feedback - I've had to try and reverse engineer this.. which was an _interesting_ experience.

Tasks for my list:
 - [x] Build the JSON
 - [x] Double-check with the Responses
 - [x] Validate using the extension
 - [x] Generate a client to check it works as expected

---

This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process. 

### PR information
- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [x] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [x] If applicable, the PR references the bug/issue that it fixes.
- [x] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [x] My spec meets the review criteria:
  - [x] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [x] Validation errors from the [Linter extension for VS Code](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/linter.md) have all been fixed for this spec. (**Note:** for large, previously checked in specs, there will likely be many errors shown. Please contact our team so we can set a timeframe for fixing these errors if your PR is not going to address them).
  - [x] The spec follows the patterns described in the [Swagger good patterns](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-good-patterns.md) document unless the service API makes this impossible.

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@tombuildsstuff__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.azure.com.
    </span>

TTYL, AZPRBOT;

File: arm-insights/2014-04-01/swagger/insights.json [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+                            }
+                        }
+                    }
+                },
+                "x-ms-pageable": {
+                    "nextLinkName": "nextLink"
+                },
+                "x-ms-odata": "#/definitions/Resource"
+            },
+            "post": {
```
Comment (tombuildsstuff) [inline arm-insights/2014-04-01/swagger/insights.json]: put

File: arm-insights/2014-04-01/swagger/insights.json [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+                "description": "Creates or Updates an Application Insights instance",
+                "operationId": "Insights_CreateUpdate",
```
Comment (tombuildsstuff) [inline arm-insights/2014-04-01/swagger/insights.json]: CreateOrUpdate to match the others

File: arm-insights/2014-04-01/swagger/insights.json [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+                    "readOnly": true,
+                    "description": "Azure resource type"
+                },
+                "location": {
+                    "type": "string",
+                    "description": "Resource location"
+                },
+                "kind": {
+                    "type": "string",
+                    "description": "Application Insights Kind"
```
Comment (tombuildsstuff) [inline arm-insights/2014-04-01/swagger/insights.json]: this wants to be the enum too

File: arm-insights/2014-04-01/swagger/insights.json [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+        "Properties": {
+            "properties": {
+                "ver": {
+                    "type": "string",
+                    "readOnly": true,
+                    "description": "Application Insights Version"
+                },
+                "applicationId": {
+                    "type": "string",
+                    "description": "XXX"
```
Comment (tombuildsstuff) [inline arm-insights/2014-04-01/swagger/insights.json]: provided by the user, needs to be unique

File: arm-insights/2014-04-01/swagger/insights.json [hunk 5]
```diff
@@ -0,0 +1,342 @@
+{
```
Comment (amarzavery) [inline arm-insights/2014-04-01/swagger/insights.json]: indentation:  2 spaces please.

**Comment by amarzavery** [comment]
@gucalder - Since you are from the insights team, can you please review this PR? 
@tombuildsstuff - We are following a [composite model](https://github.com/Azure/azure-rest-api-specs/blob/master/arm-insights/compositeInsightsManagementClient.json) for insights. 

Looks like these APIs are not covered in the current specs that are already merged in the master branch.

**Comment by tombuildsstuff** [comment]
@amarzavery as I'm new to this - I'm guessing that means I just add the file path to that file?

**Comment by tombuildsstuff** [comment]
Is there a command line linter I should be using for this? I'm using Atom and the Swagger Editor to edit this, so I assumed it'd be the default settings - I dont want to install VSCode - so the plugin there is a nonstarter for me :)

File: arm-insights/2014-04-01/swagger/insights.json (hunk 5 above)
Comment (tombuildsstuff) [inline arm-insights/2014-04-01/swagger/insights.json]: updated

**Comment by tombuildsstuff** [comment]
@amarzavery updated this to have an entry in the composite spec / be named descriptively

**Comment by tombuildsstuff** [comment]
This has been open for nearly 2 weeks with no responses

inline

could someone take a look please? :)

File: arm-insights/2014-04-01/swagger/insightsManagementClient_Components.json [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+          },
+          {
+            "$ref": "#/parameters/ApiVersionParameter"
+          },
+          {
+            "$ref": "#/parameters/SubscriptionIdParameter"
+          },
+          {
+            "$ref": "#/parameters/NameParameter"
+          }
```
Comment (amarzavery) [inline arm-insights/2014-04-01/swagger/insightsManagementClient_Components.json]: It is surprising to me that the request does not have a body parameter. If you were to update some properties of the ApplicationInsights instance then what would you update and how would you update?

**Comment by amarzavery** [comment]
@tombuildsstuff - Sorry for the delay in getting this merged. Will give this a higher priority amongst other items.

I was looking online to find the documentation of the REST API and I found this [link](https://msdn.microsoft.com/en-us/library/azure/dn931961.aspx). I wanted to confirm whether you were referring this doc while designing the swagger spec. 

I have been talking to teams internally and they are saying that this spec actually belongs to a different service named application-insights and not Insights. Hence this should reside in a different folder named arm-applicationInsights/<api-version>/swagger/applicationInsights.json. 

Please do not make the change yet, I am still talking to them to get more clarity on this. There are lot of changes that need to be made to this spec. I have made the changes locally. Please take a look at the attached file to see the difference. 
[ai.json.txt](https://github.com/Azure/azure-rest-api-specs/files/635132/ai.json.txt)

File: arm-insights/2014-04-01/swagger/insightsManagementClient_Components.json (hunk 6 above)
Comment (tombuildsstuff) [inline arm-insights/2014-04-01/swagger/insightsManagementClient_Components.json]: Perhaps I'm mis-understanding / you've posted this comment in the wrong place in the file - but why would the DELETE HTTP Call need a body to update?

With regards for Updating Redis instances - we're doing this via the CreateOrUpdate method - see [here for the implementation within Terraform](https://github.com/hashicorp/terraform/pull/10326/files#diff-f82489832a3be9189bfe0f08df4d2089R88) - this is hooked up and working (generated from before it was turned into a composite swagger)

**Comment by tombuildsstuff** [comment]
@amarzavery I ended up reverse-engineering the API, so I wasn't aware of the docs.

**Comment by tombuildsstuff** [comment]
@amarzavery 👋 did you hear back from the team by any chance?

I don't wish to be rude - but given this has been open for 3 weeks now (and doesn't look like it's going anywhere anytime soon) - am I better off just implementing support directly in Go? :(

**Comment by salameer** [comment]
Hello @tombuildsstuff, we would really like to apologize for the delay on response from our side, and we're currently actively looking for the right party that'll be able to review this spec and provide feedback if any before making a decision on the merge.

btw your not  being rude at all :) and we really appreciate all of your efforts and contributions on this repo and others!

**Comment by SergeyKanzhelev** [comment]
Some feedback: 
1. Need to describe pricing parameters when creating a component.
2. Do not call it application insights instance, call component everywhere

See some descriptions in here that can be re-used: https://docs.microsoft.com/en-us/azure/application-insights/app-insights-powershell#create-an-azure-resource-manager-template

**Comment by d3r3kk** [comment]
Hello all, I'm from the Application Insights team and I've been asked to look into this issue and to help provide resolution to this PR. 

Thanks so much @tombuildsstuff for providing the initial work and @amarzavery for adding your changes. I will likely have questions for you two as I gain insight on this PR. Also, I'll likely touch base with @SergeyKanzhelev and @salameer too just to make sure I cover all the bases!

Not 100% certain how much time I'll need to ramp up, but I feel it should take me until next week to drive us to the solution we'll adopt going forward. Please feel free to reach out to me if you have any concerns or questions!

**Comment by d3r3kk** [comment]
Unfortunately higher priority work has gotten in my way, and has hampered me from spending the time I need to get this work done in the original time I planned. This swagger specification is the next thing on my plate however, and will remain so until my highest priority item resolves. 

_I won't be able to supply an update until the week of Feb 6th at best, from what I can tell so far._

Again, please feel free to send me messages/emails if you wish to discuss this effort further.

**Comment by salameer** [comment]
Closing this PR since we have PR #988 opened by the application insights teams. @tombuildsstuff  please let me know if that's good or reopen otherwise :)

**Comment by tombuildsstuff** [comment]
thanks @salameer

**Pull request closed without merging**